File: otaku_back/otaku_back/database/schemas/title.py
from typing import Annotated, List, Optional
from beanie import Document, Indexed

from .demographic import Demographic
from .genre import Genre
from .producer import Producer


# We'll follow the same structure as Jikan API
class Title(Document):
    mal_id: Annotated[int, Indexed(unique=True)]

    title: str
    title_english: str
    title_japanese: str
    title_synonyms: List[str]
    # titles: List[dict]

    type: str
    episodes: int
    genres: List["Genre"]
    demographics: List["Demographic"]
    synopsis: str

    # aired: dict
    # airing: bool
    status: str
    # duration: str
    producers: List["Producer"]
    year: int

    popularity: Optional[float] = None

    score: float
    rank: int

    images: dict

    class Settings:
        collection = None

    def __str__(self):
        return f'{self.title} ({self.year}, ID: {self.mal_id})'


class Anime(Title):
    class Settings:
        name = "anime"


class Manga(Title):
    class Settings:
        name = "manga"

File: ponzu-back/src/models/title.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use crate::models::genre::Genre;
use crate::models::producer::Producer;

#[derive(PartialEq, Debug, Serialize, Deserialize)]
struct Title {
    mal_id: i32,
    title: String,
    title_english: String,
    title_japanese: String,
    title_synonyms: Vec<String>,
    r#type: String,
    episodes: i32,
    genres: Vec<Genre>,
    demographics: Vec<Genre>,
    synopsis: String,
    status: String,
    producers: Vec<Producer>,
    year: i32,
    popularity: Option<f32>,
    score: f32,
    rank: i32,
    images: Value,
}

// Inherit Title struct for Anime
#[derive(PartialEq, Debug, Serialize, Deserialize)]
pub struct Anime {
    #[serde(flatten)]
    title: Title
}

#[derive(PartialEq, Debug, Serialize, Deserialize,)]
pub struct Manga {
    #[serde(flatten)]
    title: Title,
}
